### PR TITLE
Add validators for webhook

### LINKED ISSUE #13: Implment first 2 validations
1. The Kind and APIGroup specified in https://github.com/openservicemesh/osm/blob/main/charts/osm/crds/policy.yaml#L100-L109 will also be validated to ensure only kind == HTTPRouteGroup && apiGroup == specs.smi-spec.io/v1alpha4 are valid for the matches

2. validate teh duration property https://github.com/openservicemesh/osm/blob/main/charts/osm/crds/meshconfig.yaml#L184

### PR DESCRIPTION
**Pull request by qnuw**
**Description**:
Add first 2 validators for webhook.

**Affected area**:

- New Functionality      [x]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

File: pkg/webhook/validators.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+func DurationValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	config := &cv1alpha1.MeshConfig{}
+	if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(config); err != nil {
+		return nil, err
+	}
+
+	d, err := time.ParseDuration(config.Spec.Certificate.ServiceCertValidityDuration)
+	if err != nil {
+		return nil, fmt.Errorf("ServiceCertValidityDuration %s is not valid", d)
```
Comment (qnuw) [inline pkg/webhook/validators.go]: @steeling Should I print d in the error or config.Spec.Certificate.ServiceCertValidityDuration?

File: pkg/webhook/validators.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+func init() {
+	gvk := metav1.GroupVersionKind{
+		Kind:    "Egress",
+		Group:   "policy.osm.io",
+		Version: "v1alpha1",
+	}
+	RegisterValidator(gvk.String(), EgressValidator)
+	RegisterValidator(gvk.String(), DurationValidator)
```
Comment (steeling) [inline pkg/webhook/validators.go]: This should be the MeshConfig gvk

File: pkg/webhook/validators.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+func EgressValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	egress := &pv1alpha1.Egress{}
+	if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(egress); err != nil {
+		return nil, err
+	}
+
+	for _, m := range egress.Spec.Matches {
+		if m.Kind != "HTTPRouteGroup" {
+			return nil, fmt.Errorf("Egress spec matches kind is %s when it should be HTTPRouteGroup", m.Kind)
```
Comment (steeling) [inline pkg/webhook/validators.go]: nit: a little less verbose, so maybe rename error message to:

"Expected Matches.Kind to be 'HTTPRouteGroup', got: %s"

and same phrasing for below

File: pkg/webhook/validators.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+		if *m.APIGroup != "specs.smi-spec.io/v1alpha4" {
+			return nil, fmt.Errorf("Egress spec matches APIGroup is %s when it should be specs.smi-spec.io/v1alpha4", *m.APIGroup)
+		}
+	}
+
+	return nil, nil
+}
+
+func DurationValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
```
Comment (steeling) [inline pkg/webhook/validators.go]: nit: I'd call this the MeshConfigValidator

File: pkg/webhook/validators.go (hunk 1 above)
Comment (steeling) [reply pkg/webhook/validators.go]: d would not have been set, so let's print the config.Spec.Certifcate...

File: pkg/webhook/validators.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	config := &cv1alpha1.MeshConfig{}
+	if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(config); err != nil {
+		return nil, err
+	}
+
+	d, err := time.ParseDuration(config.Spec.Certificate.ServiceCertValidityDuration)
+	if err != nil {
+		return nil, fmt.Errorf("ServiceCertValidityDuration %s is not valid", d)
+	}
+
```
Comment (steeling) [inline pkg/webhook/validators.go:65]: Let's add one more check, to see if D is a min duration.

At the top of the file you can add a `const MinDuration = time.Second`

This ensure they're at least not setting something negative

Comment (qnuw) [reply pkg/webhook/validators.go:65]: Added check for d < MinDuration.

File: pkg/webhook/validators.go (hunk 4 above)
Comment (qnuw) [reply pkg/webhook/validators.go]: I made 2 gvks. One for egress and another for meshconfig. Does that look right?

File: pkg/webhook/validators.go (hunk 2 above)
Comment (qnuw) [reply pkg/webhook/validators.go]: I made 2 gvks. One for egress and another for meshconfig. Does that look right?

File: pkg/webhook/validators.go (hunk 3 above)
Comment (qnuw) [reply pkg/webhook/validators.go]: Updated the error message.

File: pkg/webhook/validators.go (hunk 4 above)
Comment (qnuw) [reply pkg/webhook/validators.go]: Renamed

**Review by steeling — APPROVED** [review]

**Review by bnookala — APPROVED** [review]

**Pull request merged**